Fix swapping of tensor_content when loading a SavedModel on s390x

kun-lu20 opened the pull request:
Based on PR https://github.com/tensorflow/tensorflow/pull/45339, this PR addresses swapping of `tensor_content` data when importing a TF 2.x `SavedModel` into a TF 1.x-style `graph` on Big-Endian systems.

Test case `//tensorflow/python/keras/integration_test:saved_model_test` fails on Big-Endian systems because it tries to load a TF 2.x `SavedModel` whose `tensor_content` data had been transformed to and stored as Little-Endian format.

`load_graph()` function in `loader_impl.py` has been modified to fix this issue by swapping of `tensor_content` data from LE to BE after `meta_graph_def` proto is loaded on BE systems.

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Fix importing a SavedModel into a TF 1.x-style graph on s390x

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @k-w-w , could you please take a look at this PR? Thank you!

kun-lu20 committed:
Fix incorrect dim data when loading saved_model file on s390x

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Refactor code to remove duplication

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Update for this PR:

Test case `//tensorflow/compiler/mlir/tensorflow/tests/tf_saved_model:keras.py.test` would fail on Big-Endian systems. This test case saves a `Keras` model at first, then loads it from `saved_model` file. It uses `python` code to save the model, which swaps the tensor content for constant operators on Big-Endian systems. But when it comes to the model loading, the `c++` code doesn’t swap the tensor data, so the dimension info of the tensors becomes incorrect.

Data swap code is added in function `SavedModelV2Bundle::Load()` in `tensorflow/cc/saved_model/bundle_v2.cc` after it loads the `meta_graph_def` from `saved_model` file. Function `ByteSwapBuffer()` is added to `tensorflow/core/util/tensor_bundle/byte_swap.cc` to do the actual data swap. 

This code change would not cause any regressions on existing test cases.

kun-lu20 committed:
Resolve conflicts in reader.cc

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Merge branch 'master' into tf-s390x-loader_impl-fix

kun-lu20 commented:
Hi @ccrusius , could you please take a look at this PR? Thank you!

kun-lu20 commented:
Hi @gbaned @ccrusius , could you please have a look at the code changes? I am looking forward to hearing the comments from you. Thank you very much!

gbaned commented:
@ccrusius Can you please review this PR ? Thanks!

kun-lu20 committed:
Merge branch 'master' into tf-s390x-loader_impl-fix

kun-lu20 commented:
Hi @ccrusius, could you please take a look at this PR? Thank you very much!

gbaned commented:
@kun-lu20  Can you please approve this PR by click on Approve and run button.  Thank you.

gbaned commented:
> Hi @gbaned , I didn’t find the Approve and run button. It looks like I need a maintainer to approve running the workflow. Thanks!

Hi @kun-lu20  Sorry, I have clicked on Approve and run button. Thank you.

kun-lu20 commented:
@gbaned No problem. Thank you very much!

kun-lu20 commented:
Hi @ccrusius @gbaned , could you please review this PR? I am looking forward to hearing the feedback from you. Thank you very much!

kun-lu20 commented:
Hi @ccrusius @gbaned , could you please take a look at this PR? Thank you!

ccrusius commented:
Re-assigning to @pcish - I remember already trying to get this in but running into some test issues.

pcish commented on tensorflow/core/util/tensor_bundle/byte_swap.h:
```diff
@@ -108,6 +109,19 @@ Status ByteSwapArray(char *array, size_t bytes_per_elem, int array_len);
+Status ByteSwapBuffer(void* buff, size_t size, DataType dtype, int num_of_elem);
+
+// Swap tensor_content field of Const Op Tensors in the named functions
+Status SwapTensorContent(MetaGraphDef* meta_graph_def);
```
Again I realize you just copied the old function name, but now that this is exported by a header file, can you rename the function to ByteSwapTensorContent for consistency?

kun-lu20 replied:
Sure, the function has been renamed to `ByteSwapTensorContent`.

pcish commented on tensorflow/core/util/tensor_bundle/byte_swap.h:
```diff
@@ -108,6 +109,19 @@ Status ByteSwapArray(char *array, size_t bytes_per_elem, int array_len);
+//  dtype: type of data in this buffer.
+//  num_of_elem: number of data in this buffer, set to -1 if unspecified
+// Returns: Status::OK() on success, -1 otherwise
+Status ByteSwapBuffer(void* buff, size_t size, DataType dtype, int num_of_elem);
```
I don't believe ByteSwapBuffer will be useful outside of byte_swap.cc, can you omit this declaration in the header file and place the function definition in an unnamed namespace in byte_swap.cc? https://google.github.io/styleguide/cppguide.html#Internal_Linkage

kun-lu20 replied:
OK, I've placed the function definition in an unnamed namespace in byte_swap.cc.

pcish commented on tensorflow/core/util/tensor_bundle/byte_swap.h:
```diff
@@ -108,6 +109,19 @@ Status ByteSwapArray(char *array, size_t bytes_per_elem, int array_len);
+//  buff: pointer to the buffer to be modified IN PLACE.
+//  size: size of bytes in this buffer.
+//  dtype: type of data in this buffer.
+//  num_of_elem: number of data in this buffer, set to -1 if unspecified
```
Can you more clearly explain the meaning of the special case of passing -1 for `num_of_elem` please? When should we use -1, and what does the function do if it gets passed -1?

kun-lu20 replied:
Yes, I've updated the comments in the code. `num_of_elem` will be passed as -1 if it could not be obtained directly from tensor data. If arg `num_of_elem` is -1, function `ByteSwapBuffer()` will calculate the number of data based on arg `size` and `dtype`.

kun-lu20 committed:
Updates as per the comments from community

1. Add unamed namespace for function ByteSwapBuffer()
2. Explain the usage of arg num_of_elem in function ByteSwapBuffer()
3. Fix format issues
4. Replace function name SwapTensorContent with ByteSwapTensorContent

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented on tensorflow/core/util/tensor_bundle/byte_swap.cc:
```diff
@@ -69,6 +69,7 @@ Status ByteSwapTensor(Tensor* t) {
     case DT_UINT16:
     case DT_INT16:
       bytes_per_elem = 2;
+      array_len = (array_len == -1) ? size/bytes_per_elem : array_len;
```
Thanks! These format issues are fixed by running `clang-format`.

kun-lu20 commented on tensorflow/core/util/tensor_bundle/byte_swap.cc:
```diff
@@ -102,17 +107,59 @@ Status ByteSwapTensor(Tensor* t) {
   }
 
-  char* backing_buffer = const_cast<char*>((t->tensor_data().data()));
+  char* backing_buffer = buff;
```
No, I've removed this line from the code. Thanks.

kun-lu20 commented on tensorflow/core/util/tensor_bundle/byte_swap.cc:
```diff
@@ -102,17 +107,59 @@ Status ByteSwapTensor(Tensor* t) {
+}
+
+Status SwapTensorContent(MetaGraphDef* meta_graph_def) {
+  GraphDef graph_def = *meta_graph_def->mutable_graph_def();
```
Sure, I've removed this line.

kun-lu20 commented on tensorflow/core/util/tensor_bundle/byte_swap.cc:
```diff
@@ -102,17 +107,59 @@ Status ByteSwapTensor(Tensor* t) {
+                (*node.mutable_attr())["value"].mutable_tensor()->set_tensor_content(
+                    string(reinterpret_cast<const char*>(parsed.tensor_data().data()),
+                          parsed.tensor_data().size()));
+              }else{
```
Sure, format issues are fixed by running `clang-format`.

kun-lu20 commented on tensorflow/core/util/tensor_bundle/byte_swap.cc:
```diff
@@ -102,17 +107,59 @@ Status ByteSwapTensor(Tensor* t) {
+          if (node_value.has_tensor()) {
+            auto tsize = node_value.mutable_tensor()->tensor_content().size();
+            auto p_type = node_value.mutable_tensor()->dtype();
+            // Swap only when there is something in tensor_content field
```
I think this comment still makes sense since the swapping will happen only when `tsize != 0`.

The else branch will be run only when `tsize != 0` as well, it matches `if (parsed.tensor_data().size() != 0)`.

pcish replied:
My bad, I was matching the else to the outer if

pcish reviewed (commented):
Code looks good to me but Windows tests are failing, I'll have to check what's going on.

kun-lu20 commented:
@pcish Thanks! Since the code changes are specific to big-endian machines, it's strange that some Windows tests are failing.

pcish commented:
Update: I'm fairly certain the current Windows failures are not related. I'll keep an eye on the build status and approve this as soon as the issue(s) with the Windows build are fixed.

kun-lu20 commented:
@pcish Thank you so much!

pcish reviewed (approved):
Thanks for the fix @kun-lu20 !

kun-lu20 commented:
Thank you @pcish !
